misk-core/src/main/kotlin/misk/queuing/TimedBlockingQueue.kt

```diff
@@ -0,0 +1,169 @@
+package misk.queuing
+
+import java.time.Duration
+import java.util.concurrent.ArrayBlockingQueue
+import java.util.concurrent.BlockingQueue
+import java.util.concurrent.TimeUnit
+
+/**
+ * This BlockingQueue implementation allows measuring how long dequeued items have been
+ * in the queue. It requires a delayHandler which consumes the latency every time an item is
+ * removed from the queue.
+ * Note that the the actual removal of an item from the queue and the delayHandler invocation
+ * are not atomic, hence it may not be appropriate for applications which require
+ * a higher level of accuracy.
+ * @param <T> The type of the elements held in this queue
+ */
+class TimedBlockingQueue<T>(
+  maxQueueSize: Int,
+  private val delayHandler: (Duration) -> Unit
+) : BlockingQueue<T> {
+  private val queue: BlockingQueue<TimedQueueItem<T>> =
+      ArrayBlockingQueue<TimedQueueItem<T>>(maxQueueSize)
+
+  override fun poll(timeout: Long, unit: TimeUnit): T? {
+    val item = queue.poll(timeout, unit)
+    return if (item != null) handleRemoved(item) else null
+  }
+
+  override fun poll(): T? {
+    val item = queue.poll()
+    return if (item != null) handleRemoved(item) else null
+  }
+
+  override fun take(): T {
+    return handleRemoved(queue.take())
+  }
+
+  override fun remove(): T {
+    return handleRemoved(queue.remove())
+  }
+
+  override fun contains(element: T): Boolean {
+    return queue.contains(wrap(element))
+  }
+
+  override fun addAll(elements: Collection<T>): Boolean {
+    return queue.addAll(wrapCollection(elements))
+  }
+
+  override fun clear() {
+    invokeDelayHandlerOnAll(queue)
+    queue.clear()
+  }
+
+  override fun element(): T {
+    return unwrap(queue.element())
+  }
+
+  override fun removeAll(elements: Collection<T>): Boolean {
+    val numRemoved = removeItems(elements)
+    return numRemoved > 0
+  }
+
+  override fun add(element: T): Boolean {
+    return queue.add(wrap(element))
+  }
+
+  override fun offer(e: T): Boolean {
+    return queue.offer(wrap(e))
+  }
+
+  override fun offer(e: T, timeout: Long, unit: TimeUnit): Boolean {
+    return queue.offer(wrap(e), timeout, unit)
+  }
+
+  override fun iterator(): MutableIterator<T> {
+    return queue.map(this::unwrap).toMutableList().iterator()
+  }
+
+  override fun peek(): T? {
+    val item = queue.peek()
+    if (item != null) {
+      return unwrap(item)
+    }
+    return null
+  }
+
+  override fun put(e: T) {
+    queue.put(wrap(e))
+  }
+
+  override fun isEmpty(): Boolean {
+    return queue.isEmpty()
+  }
+
+  override fun remove(element: T): Boolean {
+    return removeItems(listOf(element)) > 0
```

Comment:
nit: be consistent when you inline calls to `removeItems`

Reply:
Are you referring to this:
```
  override fun removeAll(elements: Collection<T>): Boolean {
    val numRemoved = removeItems(elements)
    return numRemoved > 0
  }
```

Reply:
yes. sorry I just made the comment as I was scrolling through the code and felt it wasn't consistent.